### PR TITLE
contrib/labstack/echo.v4: add option to ignore errors

### PR DESCRIPTION
**Pull request by mrkagelui**
### What does this PR do?

add option to ignore errors in tracing for echo framework

### Motivation

I want to control which error should be tagged in spans and which should not

### Describe how to test/QA your changes

unit tests should be able to cover this change

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

**Comment by mrkagelui** [comment]
@gbbr sorry, I'm just tagging a random internal member. do you think I can get any review of this?

**Review by katiehockman — COMMENTED** [review]
Hello! Would you please file an issue describing your precise use case so we know what problem you're running into?
Based on your issue, we'd like to hear whether or not https://github.com/DataDog/dd-trace-go/issues/865 could meet your needs.

**Comment by mrkagelui** [comment]
> Hello! Would you please file an issue describing your precise use case so we know what problem you're running into? Based on your issue, we'd like to hear whether or not #865 could meet your needs.

Hi thanks for replying! primarily for me this is to control the status code going in error-tagged spans, but I think it's also useful in other areas such as controlling what types of custom errors should/shouldn't be tagged. also there doesn't seem to be much progress on that issue, I'd like this to be handled separately.

File: contrib/labstack/echo.v4/echotrace_test.go
Comment (rarguelloF) [inline contrib/labstack/echo.v4/echotrace_test.go:505]: I think it would probably make more sense if this is provided in each test case input (adding a new field `errorCheck` in the test case struct)

File: contrib/labstack/echo.v4/option.go
Comment (rarguelloF) [inline contrib/labstack/echo.v4/option.go:21]:
```suggestion
	errCheck      func(err error) bool
```

(to stay consistent with other contribs)

**Review by rarguelloF — CHANGES_REQUESTED** [review]
@mrkagelui thanks for your PR! the feature looks good, just some minor nits to keep the option consistent with other contribs.

**Comment by mrkagelui** [comment]
done @rarguelloF

**Comment by mrkagelui** [comment]
By the way, if it's not for consistency reasons, I'd rather name it `ignoreError`, because it's clearer what the return of function means (ignore if true), `errCheck` is very ambiguous.

**Comment by mrkagelui** [comment]
Be great if we can have some review/approval before it gets outdated again 😀 @rarguelloF

**Comment by mrkagelui** [comment]
thanks for running and let me look into the failed tests

File: contrib/labstack/echo.v4/echotrace.go
Comment (rarguelloF) [inline contrib/labstack/echo.v4/echotrace.go:90]:
```suggestion
			if err != nil && (cfg.errCheck == nil || cfg.errCheck(err)) {
```

Since the logic adding `tracer.WithError(err)` was already here (lines 99 and and 105).

Also, there are 2 more cases where the span could be marked as an error, with the custom errors we are creating from the response status code (lines 111 and 116). Maybe it would make sense to call this function there as well, to give control over these errors. What do you think?

File: contrib/labstack/echo.v4/option.go
Comment (rarguelloF) [inline contrib/labstack/echo.v4/option.go:114]:
```suggestion
// WithErrorCheck specifies a function fn which determines whether the passed
// error should be marked as an error. The fn is called whenever a request
// finishes with an error.
```

(to use a similar description as we are providing for this option in other contribs)

**Review by rarguelloF — COMMENTED** [review]
@mrkagelui thanks for addressing the comments from last review, it's looking good! 

I found a small issue that I missed in the last review, please take a look at my comment and let me know what you think!

> By the way, if it's not for consistency reasons, I'd rather name it ignoreError, because it's clearer what the return of function means (ignore if true), errCheck is very ambiguous.

I agree the name is probably improvable 😅 but in this case we prefer to give a consistent experience to the library users. Let's just make sure the public docs make it clear what the behaviour is (as you already did 🙂 ).

File: contrib/labstack/echo.v4/option.go
Comment (mrkagelui) [reply contrib/labstack/echo.v4/option.go:114]: but I think it's important to tell users what happens if this func returns true/false

File: contrib/labstack/echo.v4/echotrace.go
Comment (mrkagelui) [reply contrib/labstack/echo.v4/echotrace.go:90]: yeah I figured as well, but I think the logic is a bit tangled up... what do you think the priority should be? as a matter of fact, I think this `errCheck` option should supersede that. I initially thought of providing that when I started this PR, but then I think we should give user more control rather than just the status code.

Comment (mrkagelui) [reply contrib/labstack/echo.v4/echotrace.go:90]: see my tests. I was wanting the users themselves to do the `errors.As` and compare the code there, rather than only providing the code. they could even `errors.As` into a custom error.

Comment (mrkagelui) [reply contrib/labstack/echo.v4/echotrace.go:90]: as for the branches when there's no error but different code... I don't know, how would the users use that? compare the error string?

File: contrib/labstack/echo.v4/option.go
Comment (rarguelloF) [reply contrib/labstack/echo.v4/option.go:114]: hey! @mrkagelui sorry I missed this conversation.
The comments don't necessary need to align (I think clarifying what true/false means is a good idea), it's just the logic of what the boolean means flipped with the name change (`WithErrorCheck` will mark the error as error when returns `true`, and ignore it when it returns `false`).

Comment (mrkagelui) [reply contrib/labstack/echo.v4/option.go:131]: ok, so you are saying that since we call it `WithErrorCheck`, we'd mark it error when it's true, will make the change

File: contrib/labstack/echo.v4/echotrace.go
Comment (mrkagelui) [reply contrib/labstack/echo.v4/echotrace.go:89]: what do you think of this @rarguelloF ? how should we untangle this, i.e., the logic of this error check and the returned status code?

**Review by rarguelloF — APPROVED** [review]

**Comment by rarguelloF** [comment]
👋  @mrkagelui

For the sake of moving forward with the PR, I went ahead and I added the changes discussed in here: https://github.com/DataDog/dd-trace-go/pull/1567#discussion_r1311379717

It will be merged as soon as its reviewed and approved by someone from @DataDog/tracing-go

Thanks so much for your contribution and for your patience!

**Review by katiehockman — APPROVED** [review]

**Pull request merged**
